hack/e2e.go

```
@@ -91,76 +91,184 @@ func main() {
+			}
+		} else if *push {
+			if !finishRunning("push", exec.Command("./hack/e2e-internal/e2e-push.sh")) {
+				log.Fatal("Error pushing e2e cluster. Aborting.")
+			}
+		}
+	}
+
+	upResources := ""
+	if *checkLeakedResources {
+		upResources = ListResources()
 	}
 
 	success := true
-	switch {
-	case *ctlCmd != "":
+
+	if *ctlCmd != "" {
 		ctlArgs := strings.Fields(*ctlCmd)
 		os.Setenv("KUBE_CONFIG_FILE", "config-test.sh")
-		success = finishRunning("'kubectl "+*ctlCmd+"'", exec.Command(path.Join(*root, "cluster/kubectl.sh"), ctlArgs...))
-	case *test:
-		success = Test()
+		ctlSuccess := finishRunning("'kubectl "+*ctlCmd+"'", exec.Command("./cluster/kubectl.sh", ctlArgs...))
+		success = success && ctlSuccess
+	}
+
+	if *upgradeArgs != "" {
+		upgradeSuccess := UpgradeTest(*upgradeArgs)
+		success = success && upgradeSuccess
+	}
+
+	if *test {
+		if *skewTests {
+			skewSuccess := SkewTest()
+			success = success && skewSuccess
+		} else {
+			testSuccess := Test()
+			success = success && testSuccess
+		}
+	}
+
+	if *kubemark {
+		kubeSuccess := KubemarkTest()
+		success = success && kubeSuccess
 	}
 
 	if *down {
-		TearDown()
+		if !success && *dump != "" {
+			dumpClusterLogs(*dump)
+		}
+		tearSuccess := TearDown()
+		success = success && tearSuccess
+	}
+
+	if *checkLeakedResources {
+		log.Print("Sleeping for 30 seconds...") // Wait for eventually consistent listing
+		time.Sleep(30 * time.Second)
+		DiffResources(beforeResources, upResources, ListResources(), *dump)
 	}
 
 	if !success {
 		os.Exit(1)
 	}
 }
 
+func writeOrDie(dir string, name string, data string) string {
+	f, err := os.Create(filepath.Join(dir, name))
+	if err != nil {
+		log.Fatal(err)
+	}
+	if _, err := f.WriteString(data); err != nil {
+		log.Fatal(err)
+	}
+	if err := f.Close(); err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("Created file: %s", f.Name())
+	return f.Name()
+}
+
+func DiffResources(before string, clusterUp string, after string, location string) {
```

This function should probably return errors instead of `log.Fatal`ing them. It makes it more reusable, say if I wanted to detect leaked resources and then take some action other than crashing.